src/drivers/libraries/Irql.qll

```diff
@@ -23,6 +23,52 @@ class IrqlTypeDefinition extends SALAnnotation {
   string getIrqlMacroName() { result = irqlAnnotationName }
 }
 
+/**
+ * Represents a SAL annotation indicating that the parameter in
+ * question is used as part of adjusting the IRQL.
+ */
+class IrqlParameterAnnotation extends SALAnnotation {
+  string irqlAnnotationName;
+
+  IrqlParameterAnnotation() {
+    this.getMacroName().matches(["_IRQL_restores_", "_IRQL_saves_"]) and
+    irqlAnnotationName = this.getMacroName() and
+    exists(MacroInvocation mi | mi.getParentInvocation() = this)
+  }
+
+  string getIrqlMacroName() { result = irqlAnnotationName }
+}
+
+/**

```

Comment:
IrqlRestore and Irql Save descriptions sound like the same thing. I don't understand, can you rephrase?

Reply:
I'll try and rephrase a bit.  Essentially:
irql_restores: The value in this variable will be used to set the system IRQL.  (The variable is read from.)
irql_saves: The system will save the current IRQL to this variable.  (The variable is written to.)